core/src/main/scala/kafka/server/instrumentation/ProduceRequestInstrumentation.scala

```diff
@@ -0,0 +1,193 @@
+   */
+  def toTimeTakenInEachStageMessage(instrumentation: ProduceRequestInstrumentation): String = {
+    val marks = instrumentation.marks
+
+    // Enforce Finish is marked, so when sliding there is at list a full window
+    marks.get(Stage.Finish) match {
+      case None => instrumentation.markStage(Stage.Finish)
+      case _ =>
+    }
+
+    marks
+      .toSeq                          // Iterate over stages
+      .sortBy(s => (s._2, s._1.id))   // in order of marked timestamp then defined order of stage (to prevent multiple stages finish within 1ms)
+      .sliding(2)                     // Make sliding window to diff time taken in each stage
+      .map { window =>
+        val iter = window.iterator
+        val (stage1, timestamp1) = iter.next()
+        val (_, timestamp2) = iter.next()
+        // Use the next stage's timestamp mark minus this stage's timestamp mark to get time spent on current stage
+        // Last stage is on finish, no diff to check, i.e. no need to log
+        s"${stage1.toString}: ${timestamp2 - timestamp1}"
+      }
+      .mkString("(", ", ", ")")
+  }
+
+  /**
+   * Based on the config (e.g. probability funnel), logs the detail of the instrumentation
+   *
+   * @param request This is required to obtain the total time for threshold funnel
+   * @param instrumentation The instrumentation to log
+   */
+  def maybeLog(request: RequestChannel.Request,
+               instrumentation: ProduceRequestInstrumentation): Unit = {
+    val totalTimeMs = nanosToMs(time.nanoseconds() - request.startTimeNanos)
+    if (totalTimeMs < config.thresholdToLogMs) {
+      logger.debug("totalTimeMs={} is below {}={}, skip logging",
+                   totalTimeMs,
+                   KafkaConfig.LiLongTailProduceRequestLogThresholdMsProp,
+                   config.thresholdToLogMs)
+      return
+    }
+
+    val coin = rnd.nextDouble()
+    if (coin > config.logRatio) {
+      logger.debug("Funneled out by random coin {} > configured ratio {}", coin, config.logRatio)
+      return
+    }
+
+    // Details like instrumentation.topicPartitionsInRequest cannot be obtained from request object,
+    // because the underlying details, at this moment, have been cleared by ProduceRequest#clearPartitionRecords()
+    logger.info(
+      s"totalTimeMs=${totalTimeMs}; acks=${instrumentation.requiredAcks}; "
+        + s"topic_partitions=${instrumentation.topicPartitionsInRequest.map(tp => s"${tp.topic()}-${tp.partition()}").mkString("(", ", ", ")")}; "
+        + s"instrumentationEachStageMs=${toTimeTakenInEachStageMessage(instrumentation)}; "
+        // Shrinking ISR info is important to the produce request.
+        // When there are dead followers, an acks=-1 produce may block until the leader discovered it is dead **AND** the full ISR shrinkage to complete.
+        + s"${classOf[PendingShrinkIsr].getSimpleName}_count=${replicaManager.numOfPartitionOfIsrState(classOf[PendingShrinkIsr])}"
```

Comment:
Do you have a sample showing what this looks like in the end? It's hard to visualize its overall verbosity, etc.

Reply:
Included in the description.  Let me also post it here.
```
[2023-06-29 13:30:30,534] DEBUG Funneled out by random coin 0.6153694857122864 > configured ratio 0.5 (produce.request.instrumentation.logger)
[2023-06-29 13:39:38,139] INFO totalTimeMs=28.37; acks=-1; topic_partitions=(test1-0); instrumentationEachStageMs=(Init: 1, Authorization: 1, BeginAppendRecords: 1, AppendToLocalLog: 7, ProcessAppendToLocalLogStatus: 2, PrepareDelayedProduce: 1, EnqueueDelayedProduce: 11, BeginResponseCallback: 1, ResponseThrottling: 2); PendingShrinkIsr_count=0 (produce.request.instrumentation.logger)
```

Reply:
Thanks, I forgot it was in the PR description.

Does that include network time, or was it two nodes on a single host? It would be really handy to see the range of times collected on a dev or cert cluster with organic traffic (like BMM rather than pubsub-stressor), though even generated produce requests would be interesting. If Dave hasn't OSUA-upgraded the dev clusters yet, maybe you could work with him to delay the upgrade until this code is running and produce traffic is set up? (None of this is a blocker for checkin, it's just that OSUA happens only once a quarter, so this is a great opportunity if it's not already done.)

Reply:
It's 2 nodes on a single host, but the log lines belong to the leader broker (I configured to separate the log file dir for them).

Also note that, it includes the delay for propagation. As handled here, if no other replicas required to be awaited (RF or acks <= 1), in won't include the `PrepareDelayedProduce stage`.
```scala
      if (delayedProduceRequestRequired(requiredAcks, entriesPerPartition, localProduceResults)) {
        produceRequestInstrumentation.markStage(Stage.PrepareDelayedProduce)
```

Reply:
Yeah, I just thought the 11 ms was a little short for a two-part network hop + follower processing time, but maybe that's not too far off from what we'd see in prod data centers.